### PR TITLE
HPCC-24238 Unsafe code calling setenv after fork

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith Please review.

File: system/jlib/jthread.cpp [hunk 1]
```diff
@@ -2031,6 +2058,31 @@ protected: friend class PipeWriterThread;
...
+            }
+#ifdef _WIN32
+            char **e = _environ;
+#else
+#if defined (__APPLE__)
+            char **e = *_NSGetEnviron();
+#else
+            char **e = environ;
+#endif
+#endif
```
Comment (jakesmith) [inline system/jlib/jthread.cpp:2076]: can the above code and the block starting at line 38, go into platform.h , and into platform agnostic char **getEnvVars() or similar?

File: system/jlib/jthread.cpp [hunk 2]
```diff
@@ -2031,6 +2058,31 @@ protected: friend class PipeWriterThread;
...
+            while (*e)
```
Comment (jakesmith) [inline system/jlib/jthread.cpp:2077]: perhaps worth a comment here, to say that this is merging existing parent env. vars with set thread ones, with thread ones taking precedence?

File: system/jlib/jthread.cpp [hunk 3]
```diff
@@ -2085,11 +2140,10 @@ protected: friend class PipeWriterThread;
                 if (chdir(dir) == -1)
                     throw MakeStringException(-1, "CLinuxPipeProcess::run: could not change dir to %s", dir.get());
             }
-            ForEachItemIn(idx, envVars)
-            {
-                ::setenv(envVars.item(idx), envValues.item(idx), 1);
-            }
-            execvp(argv[0],argv);
+            if (envp.length())
+                execve(argv[0], argv, (char *const *) envp.detach());
```
Comment (jakesmith) [inline system/jlib/jthread.cpp:2144]: curious what happens if you didn't detach() ... ?
It would not call the StringAttr dtor in child, so would be safe anyway (?) , but it's clearer like this..

**Review by jakesmith — APPROVED** [review]
a couple of minor comments

**Comment by richardkchapman** [comment]
@ghalliday Please review

**Review by ghalliday — COMMENTED** [review]
@richardkchapman

**Comment by richardkchapman** [comment]
@ghalliday Repushed to address your comments

File: system/jlib/jthread.cpp [hunk 4]
```diff
@@ -2175,10 +2213,30 @@ protected: friend class PipeWriterThread;
...
+        else
+            env.append(s);
     }
-    
+
+    aindex_t lookupEnv(const char *s)
+    {
+        ForEachItemIn(idx, env)
+        {
+            const char *v = env.item(idx);
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jlib/jthread.cpp:2229]: Should be a common base-class, but I didn;t want to refactor to that extent. The Windows version does not support passing environment variables anyway!
